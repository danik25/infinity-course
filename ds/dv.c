#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/
#include <string.h> /*memmove*/
#include "dv.h"

struct dynamic_vector{

	void *array;
	size_t size; /* number of elements */ 
	size_t init_capacity; /* minimum capacity */
	size_t current_capacity; 
	size_t element_size;

};

dv_t *DVCreate(size_t capacity, size_t element_size)
{
    dv_t *vector = (dv_t*) malloc(sizeof(dv_t));
    vector -> array = (dv_t*) malloc(capacity * element_size);
    vector -> element_size = element_size;
    vector -> init_capacity = capacity;
    vector -> current_capacity = capacity;
    vector -> size = 0;
    
    if (NULL == vector || NULL == vector -> array)
    {
    	return NULL;
    }
    return vector;
}

void DVDestroy(dv_t *vector)
{
    assert(vector);
	
    free(vector -> array);
    vector -> array = NULL;
    free(vector);
    vector = NULL;
}

int DVPushBack(dv_t *vector, const void *element)
{
    dv_t *tmp = NULL;
    assert(vector);
	assert(element);
	if (vector -> size == vector -> current_capacity )
	{
	    tmp = realloc(vector -> array,
	                 2 * (vector -> current_capacity * vector -> element_size));
	    if (!tmp)
	    {
	        return 1;
	    }
	    vector -> array = tmp;
	    vector -> current_capacity *= 2;
	}

	memmove((char*)vector -> array + (vector -> size * vector -> element_size),
	    element, vector -> element_size);
	vector -> size += 1;
		
	return 0;
}

void *DVGetItemAddress(const dv_t *vector, size_t index)
{
    assert(vector);
    if(index >= vector -> size)
    {
        return (NULL);
    }
    return ((char*)vector -> array + index * (vector -> element_size));
}

void DVPopBack(dv_t *vector)
{
    dv_t *tmp = NULL;
    size_t tmp_vector = vector -> current_capacity;
    assert(vector);
    
    if(0 == vector -> size)
    {
        return;
    }  
    
    vector -> size -= 1;
    /* making the vector samller by half */
    if (vector -> size * 4 <= vector -> current_capacity)
    {
        if(vector -> init_capacity * 2 <= vector -> current_capacity)
        {
            tmp_vector /= 2;
        }
        else
        {
            tmp_vector = vector -> init_capacity;
        }
        
        tmp = realloc(vector -> array, 
                            tmp_vector * vector -> element_size);
        /* checking if the reallocation failed */
        if (tmp)
        {
            vector -> current_capacity = tmp_vector;
            vector -> array = tmp;
        }
    }
}

size_t DVSize(const dv_t *vector)
{
    assert(vector);
    return vector -> size;
}

size_t DVCapacity(const dv_t *vector)
{
    assert(vector);
    return vector -> current_capacity;
}

int DVReserve(dv_t *vector, size_t new_capacity)
{
    dv_t *tmp = NULL;
    assert(vector);
    if (new_capacity > vector -> current_capacity)
    {
        tmp = realloc(vector -> array, 
                        new_capacity * vector -> element_size);
        /* checking if the reallocation failed */
        if (tmp)
        {
            vector -> array = tmp;
            vector -> current_capacity = new_capacity;
            return 0;
        }
    }
    return 1;
}



