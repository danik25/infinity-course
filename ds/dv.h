/****************************************************************
*                       Data Structures - Dynamic vector        *
* Related files: dv.c, dv_test.c, dv.h                          *
* #Version: V.01                                                *
* writer: Dani Benjmain                                         *
* reviewer: Kobi Madrish                                        *
* last update: 1.4.19                                           *
****************************************************************/ 
/*
*   Dynamic vector is a sequence container representing 
*   array that can change in size.
*   Dynamic vector use contiguous storage locations for its elements,
*   and its size can change dynamically.
*/  

#ifndef _DV_H_
#define _DV_H_

#include <stdio.h> /* size_t */

typedef struct dynamic_vector dv_t;

/*  
*   dv_t *DVCreate(size_t capacity, size_t element_size)  
* 
*   Description:
*       Create a dynamic vector, initializing its contents 
*       depending on given capacity and element size. 
*
*   Params:
*       capacity:  the maximum size of the vector. [size_t]
*       element_size: size of element. [size_t]  
*    
*   Return value: 
*       the new vector. [dv_t *]
*/
dv_t *DVCreate(size_t capacity, size_t element_size);


/*  
*   void DVDestroy(dv_t *vector)
*   
*   Description:
*       Destroy the vector, and deallocate all the storage 
*       capacity allocated by the vector using its allocator.  
*
*   Params:
*       vector: the vector to destroy. [dv_t *] 
*/
void DVDestroy(dv_t *vector); 


/*  
*   int DVPushBack(dv_t *vector, const void *element) 
*  
*   Description:
*       Adds a new element at the end of the vector, 
*       after its current last element.
*    
*   Params:
*       vector: the vector to append to it. [dv_t *]
*       element: the element to be add. [Const void *]
*      
*   Return value: 
*       0 if success, 1 if fail (if the allocation 
*       request does not succeed). [int]
*/
int DVPushBack(dv_t *vector, const void *element);


/*  
*   void DVPopBack(dv_t *vector)  
* 
*   Description:
*       Removes the last element in the vector. 
*  
*   Params:
*       vector: the vector to remove from. [dv_t *]
*/
void DVPopBack(dv_t *vector);


/*  
*   void *DVGetItemAddress(const dv_t *vector, size_t index)  
*
*   NOTE: 
*       DVPushBack() and DVPopBack invalidates 
*       DVGetItemAddress()'s reurn value !!!!!
*
*   Description:
*       Get the address of the item at 
*       position <index> in the vector.
*        
*   Params:
*       vector: the vector to get the item address from it. [const dv_t *]
*       index: position of an element in the vector. [size_t]    
*  
*   Return value: 
*       the address of the item at the specified position in the vector [void *]
*/
void *DVGetItemAddress(const dv_t *vector, size_t index);


/*  
*   size_t DVSize(const dv_t *vector)
*  
*   Description:
*       Get the number of elements in the vector.
*    
*   Params:
*       vector: the vector to get its size. [const dv_t *]
*      
*   Return value: 
*       the nubmer of elements in the vector. [size_t]
*/
size_t DVSize(const dv_t *vector);


/*  
*   size_t DVCapacity(const dv_t *vector)
*  
*   Description:
*       Get size of current allocated storage capacity.
*    
*   Params:
*       vector: the vector to get its capacity. [const dv_t *]
*      
*   Return value: 
*       the size of the storage space currently allocated for 
*       the vector, expressed in terms of elements. [size_t]
*/
size_t DVCapacity(const dv_t *vector);
 
 
/*  
*   int DVReserve(dv_t *vector, size_t new_capacity)
*  
*   Description:
*       Request a change in capacity of the vector.
*
*   NOTE: If new_capacity < size, do not make any changes and return success.
*    
*   Params:
*       vector: the vector to change its capacity. [const dv_t *]
*       new_capacity: the new capacity to be set.[size_t]
*      
*   Return value: 
*       0 if success, 1 if fail(if the allocation 
*       request does not succeed). [int] 
*/
int DVReserve(dv_t *vector, size_t new_capacity);

#endif /* _DV_H_ */
