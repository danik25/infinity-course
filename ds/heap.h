/****************************************************************
*                Data Structures - Heap
* file: heap.h
* Related files: heap.c heap_test.c
* #Version: V 1.2
* Writer: Dani Benjamin
* Reviewer: Moshe Greenberg
* Last update: 06/06/19
***************************************************************/

#ifndef _HEAP_
#define _HEAP_

#include <stddef.h>  /* size_t */


/*
 * API Design: a heap is a specialized tree-based data structure which is
 *             essentially an almost complete[1] tree that satisfies the heap
 *             property: in a max heap, for any given node C, if P is a parent
 *             node of C, then the key (the value) of P is greater than or
 *             equal to the key of C.
 */



typedef struct heap heap_t;

/* compare_t  Description:
 *
 * return value: an integer "num" such that:
 *
 *  num < 0 if data1 < data2
 *  num = 0 if data1 = data2
 *  num > 0 if data1 > data2
 *
*/
typedef int (compare_t)(const void *data1, const void *data2, void *param);


/* is_match_t  Description: determine if param is equal to data in the heap.
 *
 * return value: Boolean  - TRUE is match, FALSE otherwise
 *
 *
*/
typedef int (is_match_t)(const void *data1, void *param);





/* Description: creates the heap and its handle.
 *
 * Params: CompareFun - the sorting function
 *         param - parameter for the funcion.
 *
 * Return Value: a pointer to the handle.
 *              NULL in case of failure.
*/
heap_t *HeapCreate(compare_t CompareFunc, void *param);


/* Description: frees the heap and its handle.
 *              If heap is NULL, no operation is performed.
 *
 * Params: heap - pointer to the handle.
 *
*/
void HeapDestroy(heap_t *heap);


/* Description: insert data to the heao
 *
 *
 *
 * Params: heap - pointer to the handle.
 *         data - data to ins ert.
 *
 * Return value: status - 0 - SUCCESS, different value otherwise.
 */
 int HeapInsert(heap_t *heap, const void *data);


/* Description: remove an element by param determined by user.
 *
 *
 *
 * Params: heap - pointer to the handle.
 *         IsMatch - matching function.
 *         param - parameter to matching function.
 *
 * Return value: pointer to the data removed, NULL if no removal
 *               was made.
 */
void *HeapRemove(heap_t *heap, is_match_t IsMatch, void *param);


/* Description: removes the top element of the heap.
 *
 *
 *
 * Params: heap - pointer to the handle.
 *
 */
void HeapPop(heap_t *heap);


/* Description: display the data of the top element of the heap.
 *
 *
 *
 * Params: heap - pointer to the handle.
 *
 *
 * Return value: the data of the top element - NULL if heap is empty.
 */
void *HeapPeek(const heap_t *heap);



/* Description: retrieves the number of elements in the heap.
 *
 *
 *
 * Params: heap - pointer to the handle.
 *
 *
 * Return value: number of elements.
 */
size_t HeapSize(const heap_t *heap);


/* Description: checks if the heap is empty.
 *
 *
 *
 * Params: heap - pointer to the handle.
 *
 *
 * Return value: Boolean - TRUE (1) if empty, FALSE(0) otherwise.
 */
int HeapIsEmpty(const heap_t *heap);


#endif /* _HEAP_ */