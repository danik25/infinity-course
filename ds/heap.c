/****************************************************************
*                Data Structures - Heap
* file: heap.c
* Related files: heap.h heap_test.c
* #Version: V 1.1
* Writer: Dani Benjamin
* Reviewer: Moshe Greenberg
* Last update: 06/06/19
***************************************************************/
#include <stdlib.h> /* malloc */
#include <assert.h> /* assert */

#include "heap.h"
#include "dv.h"

#define CAPACITY 10
#define PTR_SIZE sizeof(void*)
#define PARENT(index) (((index) - (1)) / (2))
#define LEFT_CHILD(index) (((2) * (index)) + (1))
#define RIGHT_CHILD(index) (((2) * (index)) + (2))
#define UNUSED(x) (void)(x)

static void SwapPointers(void **ptr1, void **ptr2);
static void HeapifyUp(void** first_address, void** second_address, 
                                            size_t index, heap_t *heap);
static void HeapifyDown(void** element, size_t index, heap_t *heap);
static int FindElementIndex(heap_t *heap, is_match_t IsMatch, void *data);
static int SmallChildIndex(heap_t *heap, size_t index); 
static int IsMatchStub(const void *data1, void *param);

enum STATUS {SUCCESS, FAIL};

struct heap
{
	size_t size;
	compare_t *CompareFunc;
	void *param;
	dv_t *vector;
};

/*****************************************************************************/

heap_t *HeapCreate(compare_t CompareFunc, void *param)
{
    heap_t* heap_handle = NULL;

    assert(CompareFunc);

    heap_handle = (heap_t*)malloc(sizeof(heap_t));

    if(NULL == heap_handle)
    {
        return(NULL);
    }

    heap_handle->size = 0;
    heap_handle->CompareFunc = CompareFunc;
    heap_handle->param = param;
    heap_handle->vector = DVCreate(CAPACITY, PTR_SIZE);

    if(NULL == heap_handle->vector)
    {
        free(heap_handle);
        heap_handle = NULL;
        return(NULL);
    }

    return(heap_handle);
}

/*****************************************************************************/

void HeapDestroy(heap_t *heap)
{
    assert(heap);

    DVDestroy(heap->vector);
    free(heap);
    heap = NULL;
}

/*****************************************************************************/

int HeapInsert(heap_t *heap, const void *data)
{
    size_t index = 0;
    void** first_address = NULL;
    void** second_address = NULL; 

    assert(heap);

    if(DVPushBack(heap->vector, &data))
    {
        return(FAIL);
    }

    index = DVSize(heap->vector) - 1;  

    first_address = DVGetItemAddress(heap->vector, index);
    second_address = DVGetItemAddress(heap->vector, PARENT(index));

    HeapifyUp(first_address, second_address, index, heap);
    ++heap->size;
    
    return(SUCCESS);
}

/*****************************************************************************/

void *HeapRemove(heap_t *heap, is_match_t IsMatch, void *param)
{
    size_t index = 0;
    void** element = NULL;
    void** parent = NULL;
    void* last_element = NULL;
    void* data = NULL;

    assert(heap);
    assert(IsMatch);

    index = FindElementIndex(heap, IsMatch, &param);

    /* if the data to remove doesn't exists */
    if(index >= heap->size)
    {
        return (NULL);
    }
    element = DVGetItemAddress(heap->vector, index);

    /* switching between the element to be removed and the last element, 
                                        and then removing the last element */
    data = *(void**)DVGetItemAddress(heap->vector, index);
    last_element = DVGetItemAddress(heap->vector, heap->size - 1);
    SwapPointers(element, last_element);
    
    --heap->size;

    if(LEFT_CHILD(index) > heap->size - 1)
    {
        parent = DVGetItemAddress(heap->vector, PARENT(index));
        HeapifyUp(element, parent, index, heap);
    }
    else
    {
        HeapifyDown(element, index, heap);
    }

    DVPopBack(heap->vector);
    return(data);
}

/*****************************************************************************/

void HeapPop(heap_t *heap)
{
    assert(heap);

    HeapRemove(heap, IsMatchStub, heap->param);
}

/*****************************************************************************/

void *HeapPeek(const heap_t *heap)
{
    void **tmp_data = NULL;
    assert(heap);

    tmp_data = DVGetItemAddress(heap->vector, 0);

    if(NULL == tmp_data)
    {
        return (NULL);
    }

    return(*tmp_data);
}

/*****************************************************************************/

size_t HeapSize(const heap_t *heap)
{
    assert(heap);

    return(heap->size);
}

/*****************************************************************************/

int HeapIsEmpty(const heap_t *heap)
{
    assert(heap);

    return(0 == heap->size);
}

/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

static void SwapPointers(void **ptr1, void **ptr2)
{
    void *temp = NULL;

    assert(ptr1);
    assert(ptr2);

    temp = *ptr1;
    *ptr1= *ptr2;
    *ptr2 = temp;
}

static void HeapifyUp(void** first_address, void** second_address, 
                                            size_t index, heap_t *heap)
{
    while(index > 0 && (heap->CompareFunc(*first_address, *second_address, 
                                                            heap->param) < 0))
    {
        SwapPointers(first_address, second_address);

        index = PARENT(index);

        first_address = DVGetItemAddress(heap->vector, index);
        second_address = DVGetItemAddress(heap->vector, PARENT(index));
    }
}

static void HeapifyDown(void** element, size_t index, heap_t *heap)
{
    void** small_child = NULL;
    size_t child_index = 0;

    child_index = SmallChildIndex(heap, index);

    small_child = DVGetItemAddress(heap->vector, child_index); 

    while(child_index < heap->size && 
            (heap->CompareFunc(*element, *small_child, heap->param) >= 0))
    {
        SwapPointers(element, small_child);

        element = small_child;

        child_index = SmallChildIndex(heap, child_index);
        small_child = DVGetItemAddress(heap->vector, child_index);
    }
}

static int SmallChildIndex(heap_t *heap, size_t index) 
{
    void** left_child = NULL;
    void** right_child = NULL;
    int child_index = 0;

    if(RIGHT_CHILD(index) > heap->size - 1 || LEFT_CHILD(index) > heap->size)
    {
        return(LEFT_CHILD(index));
    }

    left_child = DVGetItemAddress(heap->vector, LEFT_CHILD(index));
    right_child = DVGetItemAddress(heap->vector, RIGHT_CHILD(index));

    child_index = 
            (heap->CompareFunc(*left_child, *right_child, heap->param) < 0) ?
            LEFT_CHILD(index) : RIGHT_CHILD(index);

    return(child_index);
}

static int FindElementIndex(heap_t *heap, is_match_t IsMatch, void *data)
{
    size_t i = 0;
    void **elem = NULL;
    void **p_data = data;

    for(i = 0; i < heap->size; ++i)
    {
        elem = DVGetItemAddress(heap->vector, i);
        if(IsMatch(*elem, *p_data))
        {
            return(i);
        }
    }
    return(i);
}

static int IsMatchStub(const void *data1, void *param)
{
    UNUSED(data1);
    UNUSED(param);
    return(1);
}