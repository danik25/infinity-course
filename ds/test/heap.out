/home/xyz/git/dani-benjamin/ds/heap.out